### ISSUE TITLE
vfs: --vfs-cache-mode writes reads files from source even when they are in the VFS cache

### ISSUE THREAD

**Issue by thestigma** [rclone/rclone#3330]
#### What is your current rclone version (output from `rclone version`)?
rclone v1.48.0-063-gd7016866-beta
- os/arch: windows/amd64
- go version: go1.12.3
Windows 10

#### What problem are you are trying to solve?
Currently the cache backend's writes = true seems to be designed purely to provide safe upload in case retries are needed. Once files are confirmed uploaded to the destination the written files are removed from cache. (redundant for mounted drives that use mode writes).

As far as I can tell the cache can't actually do "proper" write caching with retention to quickly re-serve files that were recently written to a remote. Nor can the VFS cache apparently. That seems to be an important missing feature. Re-accessing uploaded files not long after they were uploaded I think is fairly frequent use-case for many - not to mention massively useful if you want to sync one remote to several other remotes. There should be no need to re-download such recently uploaded files if you just set up the cache to be large enough.

#### How do you think rclone should be changed to solve that?
It seems to me like this functionality could be created mostly by shortcutting existing code - hopefully making it not require much work. (perhaps naive...)

I would recommend adding a flag for "writes-retention".
When flag is enabled, simply skip the delete/invalidate step in the cache that normally happens after a confirmed successful transfer. The files then remain in cache - and there already exists functionality for serving in-cache files to the user rather than accessing the slower remote - so hopefully that part should require no changes.

This is a related issue, but I think the problem was attacked at the wrong place (temp-uploads rather than the writes=true functionality). It makes much more sense to let the cache do the caching, not a temp upload folder:
https://github.com/ncw/rclone/issues/2700

#### How I am prepared to help
I volunteer for any beta testing and bug-hunting you require.
I really don't know if I have the knowhow to mess with the code myself (as I have never used go).

**Comment by ncw** [comment]
> As far as I can tell the cache can't actually do "proper" write caching with retention to quickly re-serve files that were recently written to a remote. Nor can the VFS cache apparently.

The VFS cache should keep files around until `--vfs-cache-max-age` or `--vfs-cache-max-size` is reached so I would have thought this would work with the VFS?

**Comment by thestigma** [comment]
Well, that's what I thought at first too...

First I tried the cache backend's write mode, but as I explained that seems to delete the writes as soon as they are sent out from the cache, so that didn't do what I expected (and as a result I assumed it was implemented purely for error-safe uploads rather than cache retention).

Then I tried testing the VFS with write caching, but that also didn't seem to do what I expected. Sure, you can make the uploaded files stay in the VFS using large age and size parameters, but when you try to re-download something you uploaded a few minutes ago (via the mounted drive) then it doesn't seem to redirect that request to the files in the VFS. The request goes to the cloud drive instead (Gdrive in my case) and the files(s) are just re-downloaded even though they already exist locally. That is of course entirely redundant and very inefficient - and completely against my assumption on how it would work.

As far as I can tell the only way to avoid redownloading recently uploads right now is to manually go into the VFS folder and copy the files from there, but that is so convoluted that I assume it's not how it was intended.

I am 99% sure I already tested this thoroughly before making this issue, but if you want I would be happy to re-test and post a reproducible step-by-step of how the VFS fails to do what you say (and I thought) it should be doing. Just say the word.

Personally I would be happy with either the VFS or the cache backend having working write-caching with retention. Either would work for me. Doing it in the cache would probably be cleaner, safer (to not have unencrpyted files linger in write-cache) and more logical in terms of what modules do what. Doing it in the VFS would probably be faster and with less redundant local copying of files - and if the intention is to merge cache functionality into the VFS in the future it might make more sense to fix that then to do more work on the cache backend.

If I may ask - did you implement the VFS logic, or was that the work of another contributor?

**Comment by thestigma** [comment]
I have retested the VFS cache to be sure...

In short - the cache works as long as the file has not yet been uploaded to cloud. However, after the upload is complete I see in the log:
2019/07/26 19:24:46 INFO  : BIGFILES: put: cache expired
2019/07/26 19:24:46 INFO  : BIGFILES/BIGFILE.002: Copied (new)

I assume this means the VFS is discarding that file as part of the cache and no longer refers to it (aka junk to be eventually cleaned up).
After this, any attempt to re-download the file results in it being re-downloaded from the cloud (even though the file still exists within the VFS cache folder).

I guess that is the problem.
I don't see that I use any VFS settings that would cause this to happen, unless I am fundamentally misunderstanding something
Here is the config used for testing:

`[Teamdrive]
type = drive
client_id = [REDACTED]
client_secret = [REDACTED]
scope = drive
token = [REDACTED]
upload_cutoff = 256M
chunk_size = 256M`

`[TeamdriveCache]
type = cache
remote = Teamdrive:
chunk_size = 32M
info_age = 8760h
chunk_total_size = 10G
db_path = C:\---RCLONE_DB
chunk_path = E:\---RCLONE_CACHE
chunk_clean_interval = 5m
workers = 4
db_wait_time = 1s`

`[TeamdriveCacheCrypt]
type = crypt
remote = TeamdriveCache:/Secure
filename_encryption = standard
directory_name_encryption = true
password = [REDACTED]
password2 = [REDACTED]`

**Comment by thestigma** [comment]
I sniffed around the pull requests and saw that Remus was involved in working on the cache backend about a half year ago. I don't know if he is the main architect behind the cache as a whole, but if I understand correctly he seems to have implemented temp-upload and cache-writes?

I took the liberty of contacting him to see if he could confirm if the way I describe it functioning here is intentional or if it is actually a bug (or the result of me majorly misunderstanding something essential).

I feel like ultimately it would be best if the VFS cache could do both read and write caching on it's own. It would certainly be a lot more efficient than doing it with 2 stacked backends and avoid a lot of unnecessary local copying of files, but that seems like a much larger and more complex problem to solve than just "fixing" the write caching issue in the existing cache. I have to defer on this to the people who are actually familiar with the code and/or know what the development plans are for Rclone going forward...

**Comment by ncw** [comment]
> I sniffed around the pull requests and saw that Remus was involved in working on the cache backend about a half year ago. I don't know if he is the main architect behind the cache as a whole, but if I understand correctly he seems to have implemented temp-upload and cache-writes?

Yes @remusb implemented the cache backend

> I took the liberty of contacting him to see if he could confirm if the way I describe it functioning here is intentional or if it is actually a bug (or the result of me majorly misunderstanding something essential).
> 
> I feel like ultimately it would be best if the VFS cache could do both read and write caching on it's own. It would certainly be a lot more efficient than doing it with 2 stacked backends and avoid a lot of unnecessary local copying of files, but that seems like a much larger and more complex problem to solve than just "fixing" the write caching issue in the existing cache. I have to defer on this to the people who are actually familiar with the code and/or know what the development plans are for Rclone going forward...

I would like in the long run to unify the VFS cache and the cache backend. They do a very similar job.. The cache backend is a lot more sophisticated though, and can be used with `rclone sync` which the VFS cache can't.

**Comment by thestigma** [comment]
Well a VFS with integrated caching support for both reads and writes would be fantastic... if that's where you'd want to focus your effort (over the cache, especially when it's not your code) I'd support that decision even if it means it's going to take some time to happen.

But you made the VFS right, so could you at least confirm for me that it SHOULD be caching writes (beyond the upload completing)? Because as I said I can not get that to actually function after much testing. It looks to me like the VFS cache no longer considers a file part of it's cache after upload finishes so even if the file is still there it's "dead" and not actually used if you try to download it from the cloud again. I just don't know if this is by design, due to a bug, or due to some user-error on my part. Hope this isn't some windows quirk due to the differences in the mounting framework or something.

Solving that issue with the VFS's write caching would be nice, because the cache backend can take care of the reads.

Unfortunately no reply back from remusb - but we'll see what happens with that.

**Comment by ncw** [comment]
> Well a VFS with integrated caching support for both reads and writes would be fantastic... if that's where you'd want to focus your effort (over the cache, especially when it's not your code) I'd support that decision even if it means it's going to take some time to happen.

:-)

> But you made the VFS right, so could you at least confirm for me that it SHOULD be caching writes (beyond the upload completing)? Because as I said I can not get that to actually function after much testing. It looks to me like the VFS cache no longer considers a file part of it's cache after upload finishes so even if the file is still there it's "dead" and not actually used if you try to download it from the cloud again. I just don't know if this is by design, due to a bug, or due to some user-error on my part. Hope this isn't some windows quirk due to the differences in the mounting framework or something.

With `--vfs-cache-mode writes` it doesn't remove the cached files until the cache expires.

However (and this is the problem) if you try to open the file read only then it will stream it from the source rather than use the cached version.

If you open the file read/write you'll see something like this in the log

    2019/07/30 23:06:18 DEBUG : ls(0xc000382120): Opened existing cached copy with flags=O_WRONLY|O_APPEND

So I think to solve the issue we would need to make sure that `--vfs-cache-mode writes` looks in the cache for existing files even for just read only operations.

So in this bit of code

https://github.com/rclone/rclone/blob/c2635e39cc462210500bbcbab43147096c8cdd35/vfs/file.go#L570-L575

rclone would need to take the f.openRW() branch if the file exists in the cache.

That probably isn't too tricky...

**Comment by ncw** [comment]
Shall I take over this issue to implement that?

**Comment by thestigma** [comment]
Oh don't let me stop you :) Sounds like you know what to do here. It does seem like a simple change from the code, but I don't even have the software set up for code editing yet.

If you just make a test version I'd be happy to thoroughly beta it to make sure nothing unusual happens, although I don't really see much potential for it here.

**Comment by ncw** [comment]
OK I've ad a go at implementing that here.  I think it fixes the problem in my tests but I'd appreciate a double check - thanks!

Well done for spotting this problem - fixing it will make --vfs-cache-modes writes perform much better! It might even fix some bugs.

https://beta.rclone.org/branch/v1.48.0-103-g85557827-fix-3330-vfs-beta/ (uploaded in 15-30 mins)

**Comment by thestigma** [comment]
Exellent! After I have finished working on another troubleshooting issue that I want to confirm with 1.48 stable I will give this a thorough test and probably just use this going forward in case any rare edge-cases should crop up.

Will this be included in all "latest beta" builds going forward, or only merged at 1.49 ?

**Comment by ncw** [comment]
After you've given it a test I'll merge it into the beta then it should be in 1.49

**Comment by darthShadow** [comment]
Won't #2907 still prevent this from working properly for any case which involves renames?

**Comment by thestigma** [comment]
Sorry for taking my time with this, but I've given it a go now.
Initial testing shows some wonky behavior.

After clearing my caches (just to be safe) I've copied over
- one folder with 100 small'ish jpgs
- one folder with 2 x 1GB .bin files
I observe the upload process making sure they have transfered, then wait a couple of minutes for good measure, then try to re-download from mount.

**result:**
- Folder with small files works consistently
- Folder with large files fails (almost?) consistently and still redownloads from cloud. I think I saw one of the big files get referenced from cache once, but every other repeat of the test failed on large files.

rclone v1.48.0-103-g85557827-fix-3330-vfs-beta
- os/arch: windows/amd64
- go version: go1.12.3

If you need logs of this, please specify the flags you think you need to look at.
I will probably run some more tests on this tomorrow and see if I can spot anything spesific that triggers the buggy behavior.

**Comment by thestigma** [comment]
> 
> 
> Won't #2907 still prevent this from working properly for any case which involves renames?

I assume it will - but that seems kind of like a separate issue even if it is related. This issue only deals with the bug where rclone wasn't looking for the files in the cache in the fist place when accessed via read-only.

**Comment by ncw** [comment]
> If you need logs of this, please specify the flags you think you need to look at.

Sorry for the delay in replying.  Yes I'd love to see logs.  Not too bothered about the flags other than --vfs-cache-mode-writes.

if you could do the smallest test which fails and attach that log, it would be perfect.



> > Won't #2907 still prevent this from working properly for any case which involves renames?
> 
> I assume it will - but that seems kind of like a separate issue even if it is related. This issue only deals with the bug where rclone wasn't looking for the files in the cache in the fist place when accessed via read-only.

I'd forgotten about that issue...

When we get round to #2907 I think it should improve things.

**Comment by thestigma** [comment]
> 
> 
> > If you need logs of this, please specify the flags you think you need to look at.
> 
> Sorry for the delay in replying. Yes I'd love to see logs. Not too bothered about the flags other than --vfs-cache-mode-writes.
> 
> if you could do the smallest test which fails and attach that log, it would be perfect.
> 
> > > Won't #2907 still prevent this from working properly for any case which involves renames?
> > 
> > 
> > I assume it will - but that seems kind of like a separate issue even if it is related. This issue only deals with the bug where rclone wasn't looking for the files in the cache in the fist place when accessed via read-only.
> 
> I'd forgotten about that issue...
> 
> When we get round to #2907 I think it should improve things.

I was mostly thinking about "dump" flags. I take it that normal debug will be ok then, unless you tell me otherwise.

I will run some tests this evening (and tomorrow if needed) and try to narrow down the issue as much as I can on a dedicated test config and drive and see if it really is size related. If it is then I'm inclined to think it's some sort of timing issue rather than the actual size (ie. the large files take much longer to transfer) which maybe somehow could affect timestamps in an unintended way to cause this.

If you want to link the new code block that determines if the file is already in cache that might help me narrow down a probable cause.

**Comment by ncw** [comment]
> I take it that normal debug will be ok then, unless you tell me otherwise.

Just with `-vv` should be fine.

> If you want to link the new code block that determines if the file is already in cache that might help me narrow down a probable cause.

It is 85557827 - github normally auto links them - don't know why it didn't this time!

**Comment by ncw** [comment]
I did some testing of my own on this..

I think with big files the issue is that the timestamp may be wrong, so rclone corrects it.  This does transactions with upstream and a local MD5SUM but doesn't re-download the file.

Here is what happens on the first read

```
2019/08/23 14:00:31 DEBUG : 1GB: Open: flags=OpenReadOnly
2019/08/23 14:00:31 DEBUG : 1GB: Open: flags=O_RDONLY
2019/08/23 14:00:31 DEBUG : 1GB: >Open: fd=1GB (rw), err=<nil>
2019/08/23 14:00:31 DEBUG : 1GB: >Open: fh=&{1GB (rw)}, err=<nil>
2019/08/23 14:00:31 DEBUG : &{1GB (rw)}: Read: len=16384, offset=0
2019/08/23 14:00:31 DEBUG : 1GB: Modification times differ by 8.228971778s: 2019-08-23 12:59:41.706 +0000 UTC, 2019-08-23 13:59:49.934971778 +0100 BST
2019/08/23 14:00:33 DEBUG : 1GB: MD5 = e37115d4da0e187130ab645dee4f14ed OK
2019/08/23 14:00:33 INFO  : 1GB: Updated modification time in destination
2019/08/23 14:00:33 DEBUG : 1GB: Unchanged skipping
2019/08/23 14:00:33 DEBUG : 1GB(0xc00074c540): Opened existing cached copy with flags=O_RDONLY
```

Subsequent ones do this

```
2019/08/23 14:01:28 DEBUG : 1GB: Size and modification time the same (differ by 0s, within tolerance 1ms)
2019/08/23 14:01:28 DEBUG : 1GB: Unchanged skipping
2019/08/23 14:01:28 DEBUG : 1GB(0xc0000c9ce0): Opened existing cached copy with flags=O_RDONLY
```

So it still checks the object is up to date before opening.

I'm confident enough in this that I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.49.

If you can find problems though I'd be delighted to fix them :-)

**Comment by thestigma** [comment]
Will run some tests on the newest version and see if I can replicate the issue I saw with some proper logs. Sorry for being so sluggish in moving this forward. I've had a few too many balls in the air recently I think :)

**Comment by thestigma** [comment]
Very belated feedback - but I am cleaning up some old "backlog" ...

I am fairly confident at this point that the issues I still had were related to wrong modtime info being saved sometimes during upload via mount.

So this may still be an issue, but it is probably not an issue with the cache itself, since it is doing the right thing - just based on wrong modtime info (ie. believing the cached file does not match and thus grabbing it remotely anyway).

A lot of this was due to antivirus interference we discovered after a lot of troubleshooting.
I still think there are **some** cases where modtime is wrong (on Windows anyway). I still need to run some more extensive tests on this.

Anyway, this issue may now be a case of "barking up the wrong tree".
Let me know if you want to close this and then I can make a more targeted issue once I have the time to dig further into the potential "wrong modtime on upload via mount on Windows" issue specifically...
